nimblehq/gin-templates
# Setup migration

## What happened 👀
Resolve #28
- Add `goose` migration command to `Makefile`.
- Add .env for environment config.
- Update test workflow to install `goose`.
- Add `wait-for-postgres` script.
- Add create `healths` table migration.
- Update README.

## Insight 📝
`wait-for-postgres` is added because postgres docker container has a little bit of delay to start a database. We cannot run `goose` migrate command immediately after the container start as it will not able to connect to the database.
 
For `wait-for-postgres` bash script, I followed this [article](https://docs.docker.com/compose/startup-order/).

**Testing steps**
- Follow the step on README.
- For create project step, use this command instead to use a template from `feature/migration` branch.
  ```make
  nimble-gin -b feature/migration create
  ```
- Create the migration with the following command.
   ```make
   make migration/create MIGRATION_NAME={{migration name}}
   ```
- Migration file will be created under `database/migrations` directory.
- Add your own SQL statement to the migration file or you can use this statement.
  ```sql
  -- +goose Up
  -- +goose StatementBegin
  CREATE TABLE healths (
    id SERIAL PRIMARY KEY,
    status TEXT UNIQUE,
    created_at TIMESTAMP NOT NULL DEFAULT NOW(),
    updated_at TIMESTAMP NOT NULL DEFAULT NOW()
  );
  -- +goose StatementEnd

  -- +goose Down
  -- +goose StatementBegin
  DROP TABLE healths;
  -- +goose StatementEnd
  ```
- Start development server with `make dev` command
- Access to the database. You should see your SQL result.
- You can test the rollback by using `make db/rollback` command.

## Proof Of Work 📹
- All tests pass on local.
<img width="1259" alt="Screen Shot 2564-02-24 at 15 24 41" src="https://user-images.githubusercontent.com/29707647/108969956-6b38af80-76b4-11eb-846b-db1117561911.png">


- All tests pass on CI (I tested on the [testing repo](https://github.com/gutakk/test-gin-templates/actions/runs/595249962)).
![Screen Shot 2564-02-24 at 15 25 15](https://user-images.githubusercontent.com/29707647/108970332-81df0680-76b4-11eb-8cb9-352c1cf6afff.png)


- `healths` table is created after running the development server.
<img width="583" alt="Screen Shot 2564-02-24 at 15 26 42" src="https://user-images.githubusercontent.com/29707647/108971160-b357d200-76b4-11eb-9556-77f8aa770009.png">
